### ISSUE TITLE
The method "beforeStoppingVertx" in Launcher has not been invoked.

### ISSUE THREAD

**Issue by cyhii** [eclipse-vertx/vert.x#4291]
### Questions

I write a custom Launcher in my application, when I debug the lifecycle hooks I saw the "beforeStoppingVertx" has not been invoked. Here is my Launcher:

```java
public class MyLauncher extends Launcher {

    private static final Logger log = LoggerFactory.getLogger(MyLauncher.class);

    public static void main(String[] args) {
        new MyLauncher().dispatch(args);
    }

    public static void executeCommand(String cmd, String... args) {
        new MyLauncher().execute(cmd, args);
    }

    @Override
    public void beforeStoppingVertx(Vertx vertx) {
        log.info("beforeStoppingVertx");
    }

    @Override
    public void afterStoppingVertx() {
        log.info("afterStoppingVertx");
    }
}
```

The info log "beforeStoppingVertx" never be printed.

I made some debug and tests, I found this issue may be caused by the CloseFuture's hooks type, it's WeakHashMap:
https://github.com/eclipse-vertx/vert.x/blob/master/src/main/java/io/vertx/core/impl/CloseFuture.java#L63

The "addCloseHook" method be called successfully when application start
https://github.com/eclipse-vertx/vert.x/blob/master/src/main/java/io/vertx/core/impl/launcher/commands/RunCommand.java#L256
, but then the hook be removed so it is actually not worked.

Don't know whether I'm right or not, need help, please.

### Version

4.2.2

**Comment by tsegismont** [comment]
Can you share a small reproducer? How do you start/stop the application?

If you add:

```java
   @Override
    public void beforeStartingVertx(VertxOptions options) {
        log.info("beforeStartingVertx");
    }
```

Is anything `beforeStartingVertx` printed to the logs?

**Comment by cyhii** [comment]
> Is anything beforeStartingVertx printed to the logs?

Yes, it is. Here is part of my application logs:

```
20:10:03.156 [main] INFO  c.t.friendship.FriendshipLauncher - beforeStartingVertx
20:10:03.412 [main] INFO  c.t.friendship.FriendshipLauncher - afterStartingVertx
20:10:03.938 [vert.x-eventloop-thread-2] INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Starting...
20:10:04.104 [vert.x-eventloop-thread-2] INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Start completed.
20:10:04.692 [vert.x-eventloop-thread-0] INFO  i.v.c.i.l.c.VertxIsolatedDeployer - Succeeded in deploying verticle
20:10:04.728 [vert.x-eventloop-thread-3] INFO  c.t.friendship.component.GRPCServer - start warm up application
20:10:04.759 [vert.x-eventloop-thread-3] INFO  c.t.friendship.component.GRPCServer - warm up completed, cost 30ms
20:10:22.709 [vert.x-worker-thread-18] INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Shutdown initiated...
20:10:22.717 [vert.x-worker-thread-18] INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Shutdown completed.
20:10:22.830 [Thread-5] INFO  c.t.friendship.FriendshipLauncher - afterStoppingVertx
```

And in fact it is pretty easy to reproduce, just make a starter application from [start.vertx.io](https://start.vertx.io/), then add some logs. Everything works well if simply run, but add some Java heap variables like `-Xmx100m -Xms100m` to limit memory size, the issue appears.

I have made a demo project, [here](https://github.com/cyhii/vertx-starter-demo)

Steps: 
`./gradlew shadowJar`

`java -Xmx100m -Xms100m -jar build/libs/starter-1.0.0-SNAPSHOT-fat.jar`
(or make the heap size more smaller)

**Comment by tsegismont** [comment]
Closed in #4293
